src/librustc_mir/transform/simplify_comparison_integral.rs

```diff
@@ -0,0 +1,252 @@
+            IntTy::Isize => scalar
+                .to_machine_isize(&tcx)
+                .map_err(err_tracer)
+                .ok()
+                .map(|x| truncate(x as u128, tcx.data_layout().pointer_size)),
+            IntTy::I8 => scalar
+                .to_i8()
+                .map_err(err_tracer)
+                .ok()
+                .map(|x| truncate(x as u128, Size::from_bits(8))),
+            IntTy::I16 => scalar
+                .to_i16()
+                .map_err(err_tracer)
+                .ok()
+                .map(|x| truncate(x as u128, Size::from_bits(16))),
+            IntTy::I32 => scalar
+                .to_i32()
+                .map_err(err_tracer)
+                .ok()
+                .map(|x| truncate(x as u128, Size::from_bits(32))),
+            IntTy::I64 => scalar
+                .to_i64()
+                .map_err(err_tracer)
+                .ok()
+                .map(|x| truncate(x as u128, Size::from_bits(64))),
+            IntTy::I128 => scalar
+                .to_i128()
+                .map_err(err_tracer)
+                .ok()
+                .map(|x| truncate(x as u128, Size::from_bits(128))),
+        },
+        Uint(int) => match int {
+            UintTy::Usize => {
+                scalar.to_machine_usize(&tcx).map_err(err_tracer).ok().map(|x| x as u128)
+            }
+            UintTy::U8 => scalar.to_u8().map_err(err_tracer).ok().map(|x| x as u128),
+            UintTy::U16 => scalar.to_u16().map_err(err_tracer).ok().map(|x| x as u128),
+            UintTy::U32 => scalar.to_u32().map_err(err_tracer).ok().map(|x| x as u128),
+            UintTy::U64 => scalar.to_u64().map_err(err_tracer).ok().map(|x| x as u128),
+            UintTy::U128 => scalar.to_u128().map_err(err_tracer).ok().map(|x| x as u128),
+        },
+        _ => None,
+    }
+}
+
+struct OptimizationFinder<'a, 'tcx> {
+    body: &'a Body<'tcx>,
+}
+
+impl<'a, 'tcx> OptimizationFinder<'a, 'tcx> {
+    fn find_optimizations(&self) -> Vec<OptimizationInfo<'tcx>> {
+        self.body
+            .basic_blocks()
+            .iter_enumerated()
+            .filter_map(|(bb_idx, bb)| {
+                // find switch
+                let (place_switched_on, values, targets) = match &bb.terminator().kind {
+                    rustc_middle::mir::TerminatorKind::SwitchInt {
+                        discr, values, targets, ..
+                    } => Some((discr.place()?, values, targets)),
+                    _ => None,
+                }?;
+
+                // find the statement that assigns the place being switched on
+                bb.statements.iter().enumerate().rev().find_map(|(stmt_idx, stmt)| {
+                    match &stmt.kind {
+                        rustc_middle::mir::StatementKind::Assign(box (lhs, rhs))
+                            if *lhs == place_switched_on =>
+                        {
+                            match rhs {
+                                Rvalue::BinaryOp(op, left, right)
```

Comment:
I think you can do

```suggestion
                                Rvalue::BinaryOp(@op: BinOp::Eq | BinOp::Ne, left, right)
```

but you may need to turn on some feature gates